lib/text/compression/lzw.flow

```diff
@@ -1,57 +1,75 @@
 import string;
 import ds/list;
-import ds/tree;
+import ds/inttree;
 import text/blueprint;
 import math/bits;
+import loop;
 
 export {
-	// LZW-compress a string
-	lzw_encode(s : string) -> string;
-
-	// Decompress an LZW-encoded string
-	lzw_decode(data : string) -> string;
+/*
+see lib/text/compression/test_lzw.flow
+for JS:
+ Org.size |Enc.t |Dec.t | Enc.size |%org|?
+——————————+——————+——————+——————————+————+—
+    239358|    41|    33|     45144| 18%|V
+   2393580|   254|    44|    171680|  7%|V
+   4787160|   400|    67|    258520|  5%|V
+   7180740|   849|    76|    326628|  4%|V
+   9574320|  1178|    97|    388588|  4%|V
+  11967900|  1885|   121|    446260|  3%|V
+  14361480|  2359|   167|    498548|  3%|V
+  16755060|  3223|   160|    547056|  3%|V
+  19148640|  3800|   198|    593404|  3%|V
+  21542220|  4404|   180|    637988|  2%|V
+  23935800|  4862|   189|    680788|  2%|V
+*/
+  	compressLzw(s : string) -> string;
+	decompressLzw(data : string) -> string;
 }
 
-lzw_encode_rec(s : string, sLength : int, i : int, code : ref int, phrase : ref string, dict : ref Tree<string, int>, storeResult : () -> void) -> void {
+compressLzwRec(s : string, sLength : int, i : int, dict : ref IntTree<ref IntTree<int>>, code : ref int, phraseIdx : ref int, out : ref List<string>) -> void {
 	if (i < sLength) {
-		currChar = getCharAt(s, i);
-		newPhrase = ^phrase + currChar;
-		if (isSome(lookupTree(^dict, newPhrase))) {
-			phrase := newPhrase;
-		} else {
-			storeResult();
-			dict := setTree(^dict, newPhrase, ^code);
-			code := ^code + 1;
-			phrase := currChar;
+		phrase = switch (lookupIntTree(dict, ^phraseIdx)) {
+			Some(ph): ph;
+			None(): {
+				ph = makeIntTree();
+				dict := addOrUpdateIntTree(dict, ^phraseIdx, ph);
+				ph
+			};
+		};
+
+		c = getCharCodeAt(s, i);
+		switch (lookupIntTree(phrase, c)) {
+			Some(phi): {
+				phraseIdx := phi;
+			}
+			None(): {
+				code := ^code + 1;
```

Comment:
Any reason making code a ref? You are using it in a recursion.

Reply:
moving to ref saves 20% in time

Reply:
Wow.

Reply:
List mostly